lighthouse-plugin-publisher-ads/computed/long-tasks.js

```diff
@@ -0,0 +1,136 @@
+// @ts-ignore
+const LoadSimulator = require('lighthouse/lighthouse-core/computed/load-simulator');
+const MainThreadTasks = require('lighthouse/lighthouse-core/computed/main-thread-tasks');
+// @ts-ignore
+const makeComputedArtifact = require('lighthouse/lighthouse-core/computed/computed-artifact');
+// eslint-disable-next-line no-unused-vars
+const NetworkNode = require('lighthouse/lighthouse-core/lib/dependency-graph/network-node.js');
+const NetworkRecords = require('lighthouse/lighthouse-core/computed/network-records');
+const PageDependencyGraph = require('lighthouse/lighthouse-core/computed/page-dependency-graph');
+
+const LONG_TASK_THRESHOLD_MS = 50;
+
+/**
+ * @param {LH.Artifacts.TaskNode} task
+ * @param {Set<string>} knownScripts
+ * @return {boolean}
+ */
+function isLong(task, knownScripts) {
+  if (task.duration < LONG_TASK_THRESHOLD_MS) {
+    return false; // Short task
+  }
+  const script = getAttributableUrl(task, knownScripts);
+  if (!script) {
+    return false;
+  }
+  if (task.parent) {
+    // Only show this long task if doing so adds more information for debugging.
+    // So we hide it if it's attributed to the same script as the parent task.
+    const parentScript = getAttributableUrl(task.parent, knownScripts);
+    return script != parentScript;
+  }
+  return true;
+}
+
+/** Finds long tasks, with support for simulation. */
+class LongTasks extends ComputedMetric {
+  /**
+   * @param {LH.Trace} trace
+   * @param {LH.DevtoolsLog} devtoolsLog
+   * @param {LH.Audit.Context} context
+   * @return {Promise<BaseNode>} networkRecords
+   */
+  static async getSimulationGraph(trace, devtoolsLog, context) {
+    /** @type {NetworkNode} */
+    const documentNode =
+      // @ts-ignore Property 'request' does not appear on PageDependencyGraph
+      await PageDependencyGraph.request({trace, devtoolsLog}, context);
+    return AdLanternMetric.getOptimisticGraph(documentNode);
+  }
+
+  /**
+   * @param {LH.Trace} trace
+   * @param {LH.DevtoolsLog} devtoolsLog
+   * @param {LH.Audit.Context} context
+   * @return {Promise<LH.Artifacts.TaskNode[]>} networkRecords
+   */
+  static async computeSimulatedResult(trace, devtoolsLog, context) {
+    const graph = await this.getSimulationGraph(trace, devtoolsLog, context);
+    const simulator = await LoadSimulator.request(
+      {devtoolsLog, settings: context.settings}, context);
+    const {nodeTimings} = simulator.simulate(graph, {});
+
+    /** @type {LH.Artifacts.TaskNode[]} */ const tasks = [];
+    for (const [node, timing] of nodeTimings.entries()) {
+      if (node.type !== BaseNode.TYPES.CPU ||
+          timing.duration < LONG_TASK_THRESHOLD_MS) {
+        continue; // Not a long task
+      }
+      tasks.push({
+        event: node.event,
+        startTime: timing.startTime,
+        endTime: timing.endTime,
+        duration: timing.duration,
+        selfTime: timing.duration, // TODO: subtract child time
```

Comment:
Is this for a followup PR? It seems that it would have a pretty significant effect.

Reply:
We don't use `selfTime` anywhere now so it won't have any effect, but I do plan to address in a follow up PR to avoid any surprising behavior in the future